### ISSUE TITLE
Variables scope issues

### ISSUE THREAD

**Issue by jducro** [scssphp/scssphp#46]
I'm trying to compile bootstrap within a block

```scss
.dp-po {
  @import "vendor/bootstrap";
}
```

But I have 2 scope related issues: 
First keyframes are not keeping their parent scope :
```scss
@if $enable-transitions {
  @keyframes progress-bar-stripes {
    from { background-position: $progress-height 0; }
    to { background-position: 0 0; }
  }
}
```
`$progress-height` is defined within bootstrap but I have to redeclare outside my block

mixin are a marker in here: https://github.com/scssphp/scssphp/blob/master/src/Compiler.php#L4070

So the function jump to the `rootEnv` and skip all my bootstrap variables. The problem happen in the rfs mixin even with variable declared in the same file:
https://github.com/twbs/bootstrap/blob/master/scss/vendor/_rfs.scss#L135

I manage to compile everything properly by just adding `false &&` at the beginning of the if.

**Comment by jducro** [comment]
I actually have another issue, the grid is not properly generated.

The local `$i` is being ignored here: https://github.com/twbs/bootstrap/blob/master/scss/mixins/_grid-framework.scss#L37
So all columns end up with `width: 100%`

**Comment by dleffler** [comment]
I am still having a similar issue with the same scenario where it loses track of root level variables which throws the error:
Undefined variable $theme-colors: /bootstrap4/scss/_functions.scss on line 72, at column 3 Call Stack: #0 function theme-color /bootstrap4/scss/_variables.scss on line 172 

```
.exp-skin {
  $swatch: "cerulean"; // newui uses this swatch
  $enable-rounded: true; // default true
  $enable-shadows: false; // default false
  $enable-gradients: false; // default false
  $enable-transitions: true; // default true

  $themepath: "../../../themes/bootstrap4theme/scss";  // hack to prevent picking up theme styles

  @import "bootstrap";

  font-family: $font-family-sans-serif !important;
}

**Comment by Cerdic** [comment]
Debuging on the full bootstrap compilation is totaly impossible given the recurse calls and heaps.
We need to narrow to a simpler test case, as small as possible

**Comment by jducro** [comment]
I'll try to come with a replicable subset of it small enough to get it.

**Comment by Cerdic** [comment]
OK I can reproduce - at least one non-working case - with:

`t46i.scss` :
```
$color1: #eee;
$color2: #ddd;
$color3: #ccc;
$color4: #bbb;


@function bgcolor() {
  @return $color2;
}

@mixin myborder() {
  border:1px solid $color3;
  @content;
}

.var-in-class {
  color:$color1;
  background-color: bgcolor();
  @include myborder(){
    strong {
      color: $color4;
    }
  }
}
```

and `t46.scss`:
```
.skin {
	@import "t46i";
}
```

Compiling t46i.scss works:
```
$ ./bin/pscss t46i.scss
.var-in-class {
  color: #eee;
  background-color: #ddd;
  border: 1px solid #ccc; }
  .var-in-class strong {
    color: #bbb; }
```

While compiling t46.scss fails:
```
./bin/pscss t46.scss

Fatal error: Uncaught ScssPhp\ScssPhp\Exception\CompilerException: Undefined variable $color2: t46i.scss on line 8, at column 3
Call Stack:
#0 function bgcolor t46i.scss on line 18 in /Users/cedric/Sites/libs/scssphp/src/Compiler.php:4516
Stack trace:
#0 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(4141): ScssPhp\ScssPhp\Compiler->throwError('Undefined varia...')
#1 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(2974): ScssPhp\ScssPhp\Compiler->get('color2')
#2 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(2600): ScssPhp\ScssPhp\Compiler->reduce(Array, true)
#3 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(1776): ScssPhp\ScssPhp\Compiler->compileChild(Array, Object(ScssPhp\ScssPhp\Formatter\OutputBlock))
#4 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(4610): ScssPhp\ScssPhp\Compiler->compileChildren(Array, Object(ScssPhp\ScssPhp\Formatter\OutputBlock), 'function bgcolo...')
#5 /Users/cedric/Sites/libs/scssphp/src/Compiler.php(3037): ScssPhp\ScssPhp\Compiler->callScssFunction('bgcolor', Array, NULL)
#6 in /Users/cedric/Sites/libs/scssphp/src/Compiler.php on line 4516
```

**Comment by Cerdic** [comment]
And a minimal test case with `@function` is
```
.skin {
  $color2: #ddd;

  @function bgcolor() {
    @return $color2;
  }

  .var-in-class {
    background-color: bgcolor();
  }
}
```

and with `@mixin`, same as long as the mixin call not as content:
```
.skin {
  $color3: #ccc;

  @mixin myborder() {
    border:1px solid $color3;
    @content;
  }

  .var-in-class {
    @include myborder();
  }
}
```

**Comment by Cerdic** [comment]
I think this is now solved!
(but didn't try on a full bootstrap import)

**Comment by dleffler** [comment]
My specific issue when subclassing Twitter Bootstrap v4.x was FINALLY fixed in release v1.1.0. I was 'hacking' my local copy to store/restore variables which got lost during compilation. I've glad @stof added all those fixes and optimizations!
